lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -1541,6 +1551,8 @@ public function addToIdentityMap($entity)
         $idHash    = implode(' ', $identifier);
         $className = $classMetadata->rootEntityName;
 
+        $this->oidMap[$oid] = $entity;
```

Comment:
Instead of saving the whole document, just save a reference to it. This will also stop it from being garbage collected and uses much less memory

$this->oidMap[$oid] = &$entity;